pd/src/app.rs

```diff
@@ -348,22 +355,31 @@ impl App {
     }
 
     fn end_block(&mut self, end: EndBlock) -> response::EndBlock {
+        let epoch = Epoch::from_blockheight(end.height, self.epoch_duration)
+            .expect("must be able to calculate block epoch");
         self.pending_block
             .as_mut()
             .expect("pending_block must be Some in EndBlock")
             .lock()
             .unwrap()
-            .set_height(end.height);
+            .set_height_and_epoch(end.height, epoch);
 
-        if end.height < 0 {
-            panic!("block height should never be negative");
-        }
+        let is_epoch_boundary = Epoch::is_epoch_boundary(end.height, self.epoch_duration)
```

Comment:
non-blocking on this PR, but maybe it would be cleaner to have `Epoch::start_height(&self) -> block::Height` and `Epoch::end_height(&self) -> block::Height` instead of `Epoch::is_epoch_boundary()`?

That way, it's still possible to test if something is the boundary (`height == epoch.start_height()`, etc), but it's unambiguous which side of the boundary the test is checking, whereas with `is_epoch_boundary` it's not immediate whether that's checking for the first or last block.

Reply:
xref https://github.com/penumbra-zone/penumbra/pull/299/files#r771246246